systemd: Fix systemd-{cryptenroll,cryptsetup} TPM2 and FIDO2 support (attempt #3)

zhaofengli opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR makes it possible to use `systemd-cryptenroll` (and its integration in `systemd-cryptsetup`) to automatically unlock LUKS volumes using TPM2 or FIDO2.

Based on #171242 (attempt #<span></span>1) and #179823 (attempt #<span></span>2).

Relavent discussions about potential attack vectors: https://github.com/NixOS/nixpkgs/pull/179823#discussion_r912328289

## How to test

You can test this PR with the minimal amount of rebuilds by pulling in the updated packages and NixOS module on [my `cryptenroll-master` branch](https://github.com/zhaofengli/nixpkgs/commits/cryptenroll-master):

```nix
{ pkgs, config, ... }:
let
  nixpkgs-cryptenroll = builtins.fetchTarball {
    name = "nixpkgs-cryptenroll";
    url = "https://github.com/zhaofengli/nixpkgs/archive/4b00eacbb8bd00dc08b6ebef6d220b3d8e49ecb4.tar.gz";
    sha256 = "0776mjndy9j7bsaa077js6bsrsanb102rccnm4kf5xzyrsmrh4d9";
  };
  pkgsPatched = import nixpkgs-cryptenroll {
    system = pkgs.system;
    overlays = [];
  };

  # HACK to use the same version of tpm2-tss and libfido2
  overlaidModule = let
    mod = import (nixpkgs-cryptenroll + "/nixos/modules/system/boot/systemd/initrd.nix");
  in { lib, config, utils, pkgs, ... } @ args: mod (args // {
    pkgs = pkgs // {
      inherit (pkgsPatched) tpm2-tss libfido2;
    };
  });
in {
  disabledModules = [
    "system/boot/systemd/initrd.nix"
  ];

  imports = [
    overlaidModule
  ];

  boot.initrd.systemd = {
    enable = true;
    package = pkgsPatched.systemdStage1;
    emergencyAccess = true;
    initrdBin = with pkgs; [ gnugrep strace ]; # for debugging only
  };

  boot.initrd.luks.devices = {
    yourvolume = {
      # ... remaining configs ...

      crypttabExtraOpts = [
        #"fido2-device=auto"
        "tpm2-device=auto"
      ];
    };
  };

  boot.initrd.availableKernelModules = [ "tpm-tis" "tpm-crb" ];
}
```

Reboot into the new configuration. _After rebooting_, add a new LUKS key slot with the key sealed to the current PCR values:

```
sudo systemd-cryptenroll --tpm2-pcrs=0+2+4+8+9 --tpm2-device=auto /dev/disk/by-uuid/your-uuid-here
```

**Note: The PCR list here is an example only! You need to understand what's being measured and when. One bootloader-specific thing is initrd: https://github.com/systemd/systemd/issues/12716**

Then, reboot again. This time the volume should be decrypted without prompting you for a password.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

oxalica commented on pkgs/os-specific/linux/cryptsetup/relative-token-path.patch:
```diff
@@ -0,0 +1,18 @@
+diff --git a/lib/luks2/luks2_token.c b/lib/luks2/luks2_token.c
```
https://github.com/NixOS/nixpkgs/pull/179823#discussion_r927352181

> please make `relative-token-path.patch` `git format-patch`-style, and add the explanation in the "commit message". That way it won't need to be explained over and over again, but lives alongside the patch.

RaitoBezarius commented:
Do you think you can add a NixOS test?

JJJollyjim commented:
@RaitoBezarius I was already writing one when I looked back at this PR to check something and saw your comment… should be done in an hour or so

JJJollyjim commented:
Here we go:

```nix
import ./make-test-python.nix ({ lib, pkgs, ... }: {
  name = "systemd-initrd-luks-tpm";

  nodes.machine = { pkgs, ... }: {
    # Use systemd-boot
    virtualisation = {
      emptyDiskImages = [ 512 ];
      useBootLoader = true;
      useEFIBoot = true;
      qemu.options = ["-chardev socket,id=chrtpm,path=/tmp/mytpm1/swtpm-sock -tpmdev emulator,id=tpm0,chardev=chrtpm -device tpm-tis,tpmdev=tpm0"];
    };
    boot.loader.systemd-boot.enable = true;

    boot.initrd.availableKernelModules = [ "tpm_tis" ];

    environment.systemPackages = with pkgs; [ cryptsetup ];
    boot.initrd.systemd = {
      enable = true;
    };

    specialisation.boot-luks.configuration = {
      boot.initrd.luks.devices = lib.mkVMOverride {
        cryptroot = {
          device = "/dev/vdc";
          crypttabExtraOpts = [ "tpm2-device=auto" ];
        };
      };
      virtualisation.bootDevice = "/dev/mapper/cryptroot";
    };
  };

  testScript = ''
    import subprocess
    import os
    import time


    class Tpm:
        def __init__(self):
            os.mkdir("/tmp/mytpm1")
            self.start()

        def start(self):
            self.proc = subprocess.Popen(["${pkgs.swtpm}/bin/swtpm", "socket", "--tpmstate", "dir=/tmp/mytpm1", "--ctrl", "type=unixio,path=/tmp/mytpm1/swtpm-sock", "--log", "level=20", "--tpm2"])

        def wait_for_death_then_restart(self):
            while self.proc.poll() is None:
                print("waiting for tpm to die")
                time.sleep(1)
            assert self.proc.returncode == 0
            self.start()

    tpm = Tpm()


    # Create encrypted volume
    machine.wait_for_unit("multi-user.target")
    machine.succeed("echo -n supersecret | cryptsetup luksFormat -q --iter-time=1 /dev/vdc -")
    machine.succeed("PASSWORD=supersecret SYSTEMD_LOG_LEVEL=debug systemd-cryptenroll --tpm2-pcrs= --tpm2-device=auto /dev/vdc |& systemd-cat")

    # Boot from the encrypted disk
    machine.succeed("bootctl set-default nixos-generation-1-specialisation-boot-luks.conf")
    machine.succeed("sync")
    machine.crash()

    tpm.wait_for_death_then_restart()

    # Boot and decrypt the disk
    machine.wait_for_unit("multi-user.target")
    assert "/dev/mapper/cryptroot on / type ext4" in machine.succeed("mount")
  '';
})
```

Based on the existing systemd-initrd-luks-* family of tests – feel free to add to the PR.

JJJollyjim commented:
(it passes on master with this PR added, not much rebuilding required for such a simple config, took me about 10 minutes)

Kranzes commented:
Can we add to `storePaths` only the stuff we need fro pkgs.tpm2-tss and pkgs.libfido2?

Kranzes commented:
Here's another minor patch that will make the additional store paths optional.
```patch
diff --git a/nixos/modules/system/boot/systemd/initrd.nix b/nixos/modules/system/boot/systemd/initrd.nix
index f10a8c1313f..f0ec3db3b21 100644
--- a/nixos/modules/system/boot/systemd/initrd.nix
+++ b/nixos/modules/system/boot/systemd/initrd.nix
@@ -394,6 +394,14 @@ in {
         "${cfg.package}/lib/systemd/system-generators/systemd-hibernate-resume-generator"
         "${cfg.package}/lib/systemd/system-generators/systemd-run-generator"
 
+        # utilities needed by systemd
+        "${cfg.package.util-linux}/bin/mount"
+        "${cfg.package.util-linux}/bin/umount"
+        "${cfg.package.util-linux}/bin/sulogin"
+
+        # so NSS can look up usernames
+        "${pkgs.glibc}/lib/libnss_files.so.2"
+      ] ++ optionals cfg.package.withCryptsetup [
         # tpm2 and fido2 support
         "${cfg.package}/lib/cryptsetup/libcryptsetup-token-systemd-tpm2.so"
         "${cfg.package}/lib/cryptsetup/libcryptsetup-token-systemd-fido2.so"
@@ -402,14 +410,6 @@ in {
 
         # the unwrapped systemd-cryptsetup executable
         "${cfg.package}/lib/systemd/.systemd-cryptsetup-wrapped"
-
-        # utilities needed by systemd
-        "${cfg.package.util-linux}/bin/mount"
-        "${cfg.package.util-linux}/bin/umount"
-        "${cfg.package.util-linux}/bin/sulogin"
-
-        # so NSS can look up usernames
-        "${pkgs.glibc}/lib/libnss_files.so.2"
       ] ++ jobScripts;
 
       targets.initrd.aliases = ["default.target"];
```

zhaofengli commented:
> Can we add to `storePaths` only the stuff we need fro pkgs.tpm2-tss and pkgs.libfido2?

We could, but I'd prefer not to do that to make it less fragile and reduce work in the future in case tpm2-tss/libfido2 change their dependencies.

> Here's another minor patch that will make the additional store paths optional.

Thanks, applied.

Kranzes commented:
>We could, but I'd prefer not to do that to make it less fragile and reduce work in the future in case tpm2-tss/libfido2 change their dependencies.

Can that not be said about the rest of the stuff in `storePaths`?

zhaofengli commented:
> Can that not be said about the rest of the stuff in `storePaths`?

No, because the dependencies are listed in the ELF, unlike tpm2-tss which dlopen's libraries dynamically depending on the interface.

I went ahead and did it anyways and looks like it saved us 1MB, but at what cost?

Kranzes commented:
You're probably right, maybe just keep libfido's single library and still use pkgs.tpm2-tss?

zhaofengli commented:
To measure the impact on the initrd size, let's use the following simple expression:

```nix
let
  pkgs = import ./. {};
  nixos = pkgs.nixos {
    boot.initrd.systemd.enable = true;
  };
in nixos.config.system.build.initialRamdisk
```

<table>
    <thead>
        <tr>
            <th>tpm2-tss</th>
            <th>libfido2</th>
            <th>Size (bytes)</th>
            <th>Δ from baseline</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td colspan=2>Without added files</td>
            <td>10948957</td>
            <td>0</td>
        </tr>
        <tr>
            <td>Slim</td>
            <td>Slim</td>
            <td>13613312</td>
            <td>+2.5MiB</td>
        </tr>
        <tr>
            <td>Slim</td>
            <td>Full</td>
            <td>13760201</td>
            <td>+2.7MiB</td>
        </tr>
        <tr>
            <td>Full</td>
            <td>Slim</td>
            <td>13908566</td>
            <td>+2.8MiB</td>
        </tr>
        <tr>
            <td>Full</td>
            <td>Full</td>
            <td>14052514</td>
            <td>+3.0MiB</td>
        </tr>
    </tbody>
</table>

Note that the sizes here are after zstd compression. Including the percentage increases in the table would be misleading, because the baseline sizes in actual configs are larger due to the bundled kernel modules.

flokli commented on pkgs/os-specific/linux/cryptsetup/relative-token-path.patch:
```diff
@@ -0,0 +1,50 @@
+Date: Sun, 4 Sep 2022 11:15:02 -0600
+Subject: [PATCH] Allow loading token handlers from the default search path
+
+Since [1] landed in cryptsetup, token handlers (libcryptsetup-token-*.so)
+are loaded from a fixed path defined at compile-time. This is
+problematic with NixOS since it introduces a dependency cycle
+between cryptsetup and systemd.
+
+This downstream patch [2] allows loading token plugins from the
+default library search path. This approach is not accepted upstream [3]
+due to security concerns, but the potential attack vectors require
+root access and they are sufficiently addressed:
+
+- cryptsetup could be used as a setuid binary (not used in NixOS).
+  In this case, LD_LIBRARY_PATH is ignored because of secure-execution
+  mode.
+- cryptsetup running as root could lead to a malicious token handler
+  being loaded through LD_LIBRARY_PATH. However, fixing the path
+  doesn't prevent the same malicious .so being loaded through LD_PRELOAD.
+
+[1] https://gitlab.com/cryptsetup/cryptsetup/-/commit/5b9e98f94178d3cd179d9f6e2a0a68c7d9eb6507
+[2] https://github.com/NixOS/nixpkgs/issues/167994#issuecomment-1094249369
+[3] https://gitlab.com/cryptsetup/cryptsetup/-/issues/733
```
Thanks for writing this up :heart:

oxalica commented:
Should we remove these assertions now? Seems there are no more concerns?

https://github.com/NixOS/nixpkgs/blob/74d81b49086d812d6beb2bfaa245ba0d454dbc3f/nixos/modules/system/boot/luksroot.nix#L908-L911

zhaofengli commented:
> Should we remove these assertions now?

I reworded it since configuring through `boot.initrd.luks.devices.<name>.fido2` is still not supported.

Kranzes commented:
I recently found out about this https://github.com/bulwarkid/virtual-fido, we could maybe use this to make a test for the fido2 functionality.

ElvishJerricco commented:
After reading the comments about security in the patch file, my only concern is the [initrd size increase](https://github.com/NixOS/nixpkgs/pull/189676#issuecomment-1237463367). It's a substantial increase, whereas we've been able to keep a ~10M minimal size for a long time now. It's not *that* big, but it's certainly a fairly unprecedented increase. It makes me wonder if we need a `boot.initrd.systemd.features` option to allow configuring systemd features like `withCryptsetup` so different initrd modules can specify their systemd needs separately and so we can disable the crypto features unless they're used in the initrd luks module.

endocrimes commented on pkgs/os-specific/linux/cryptsetup/relative-token-path.patch:
```diff
@@ -0,0 +1,50 @@
+- cryptsetup could be used as a setuid binary (not used in NixOS).
+  In this case, LD_LIBRARY_PATH is ignored because of secure-execution
+  mode.
+- cryptsetup running as root could lead to a malicious token handler
```
This line should stay

zhaofengli replied:
Assuming that you mistook it for a deletion (-), this is part of the commit message and not the diff. Let me change it to `*` so it looks better in GitHub.

zhaofengli committed:
systemd/initrd: Add files required by TPM2 and FIDO2 support to the initramfs

zhaofengli committed:
nixos/luksroot: Reword message on FIDO2 support with systemd stage 1

zhaofengli committed:
systemd: Wrap in LUKS2 tokens

Update pkgs/os-specific/linux/systemd/default.nix

Co-authored-by: Janne Heß <janne@hess.ooo>
Co-authored-by: Ilan Joselevich <personal@ilanjoselevich.com>
Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

JJJollyjim committed:
nixos/tests/systemd-initrd-luks-tpm2: init

zhaofengli committed:
systemd/initrd: Add TPM modules into initrd

This improves the out-of-box experience of TPM2 unlocking at a
small (50K) overhead.

zhaofengli committed:
cryptsetup: Allow reading tokens from relative path

Co-authored-by: Janne Heß <janne@hess.ooo>
Co-authored-by: Ilan Joselevich <personal@ilanjoselevich.com>

oxalica commented:
Is there any more blockers of this? The ~3MB initrd size growth seems acceptable and could be optimized after this.
This PR is the last blocker for me to use stage 1 systemd on my laptop.

endocrimes commented:
~3MB of initrd growth for fairly large quality of life improvements seems pretty reasonable to me 👍

oxalica committed:
nixos/tests/systemd-initrd-luks-fido2: init

oxalica commented:
For anyone interested, I wrote a FIDO2 test using QEMU with CanoKey support in https://github.com/oxalica/nixpkgs/commit/d75bc7b6c040fa0774aaa8d11e3c231120ab88d5, which depends on #194254.

lovesegfault commented:
I merged https://github.com/NixOS/nixpkgs/pull/194254, this PR can be rebased and @oxalica's commit cherry-picked :)

zhaofengli commented:
Done! The test won't work in `staging` since #194254 isn't there yet, but it seems to pass on [`cryptenroll-master`](https://github.com/zhaofengli/nixpkgs/commits/cryptenroll-master).

One weird thing though, if I add `os.unlink("/tmp/canokey-file")` before the second boot the test still passes for some reason. However, the cryptenroll functionality does seem to be used. @oxalica

oxalica commented:
> One weird thing though, if I add os.unlink("/tmp/canokey-file") before the second boot the test still passes for some reason. 

Seems the private key of the virtual card is hard-coded.
https://github.com/canokeys/canokey-core/blob/56182bb4e8b0566954dff4e9af9762b26f580594/virt-card/fabrication.c

colemickens commented:
Small notes:
1. This "blocked" startup in such a way that the usual hidpi font didn't kick in until after I performed the auth.
2. I didn't see any prompt on screen, I just noticed the yubikey flashing at me.
3. By pulling this and adding a specialisation for this boot method, I ran into and fixed #194530

Otherwise it worked great, thanks!

oxalica commented:
@colemickens 
>     1. This "blocked" startup in such a way that the usual hidpi font didn't kick in until after I performed the auth.
>     2. I didn't see any prompt on screen, I just noticed the yubikey flashing at me.

I believe that's unrelated and deserves an individual issue about systemd-initrd. You could try to shuffle around the loading order of your graphics driver, between `boot.{,initrd.{available,}}kernelModules`.

This PR is not just about systemd-initrd. Currently on master, even full systemd package doesn't support FIDO2 enrolling.

arianvp commented:
@GrahamcOfBorg build cryptsetup systemd

arianvp commented:
@GrahamcOfBorg test systemd cryptsetup

arianvp commented:
@GrahamcOfBorg build cryptsetup.passthru.tests

For some reason it failed on x86? maybe flakey? just rerunning before merging

zhaofengli commented:
Rebased to resolve conflicts. The [failure](https://logs.nix.ci/logfile/nixos/nixpkgs.189676/dca5d82a-d384-48ab-b255-4cd435191024) seems to be unrelated, and the tests (`systemd`, `systemd-initrd-luks-tpm2`, `systemd-initrd-luks-fido2`) do succeed on the [cryptsetup-master](https://github.com/zhaofengli/nixpkgs/commits/4b00eacbb8bd00dc08b6ebef6d220b3d8e49ecb4) branch.

```
error: builder for '/nix/store/xpw7xrlq6gs8zm2cz071z45s6wbh1jxp-python3.10-eventlet-0.33.1.drv' failed with exit code 1;
       last 10 log lines:
       >   /build/source/eventlet/greenthread.py:221: DeprecationWarning: capitalize_response_headers is disabled.
       >    Please, make sure you know what you are doing.
       >    HTTP headers names are case-insensitive per RFC standard.
       >    Most likely, you need to fix HTTP parsing in your client software.
       >     result = function(*args, **kwargs)
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED tests/patcher_test.py::test_fork_after_monkey_patch - AssertionError: ...
       > = 1 failed, 562 passed, 120 skipped, 17 deselected, 2483 warnings in 113.06s (0:01:53) =
       For full logs, run 'nix log /nix/store/xpw7xrlq6gs8zm2cz071z45s6wbh1jxp-python3.10-eventlet-0.33.1.drv'.
```

ajs124 commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -332,7 +332,10 @@ in {
-    boot.initrd.availableKernelModules = [ "autofs4" ]; # systemd needs this for some features
+    boot.initrd.availableKernelModules = [
+      "autofs4"           # systemd needs this for some features
+      "tpm-tis" "tpm-crb" # systemd-cryptenroll
```
Can we maybe make this optional? Or conditional on if the kernel has this option set?

I have some custom kernel configs that really don't need this module and also use the new initrd and now the initrd build fails with `modprobe: FATAL: Module tpm-tis not found in directory […]`.

zhaofengli replied:
I guess we can have a new option under `boot.initrd.luks.devices.<name>` that indicates whether a LUKS makes use of cryptenroll-based TPM2/FIDO2 unlocking, and only insert the modules and extra files when a volume with those options enabled exists. However, this can be confusing for non-systemd-stage1 users, especially when `boot.initrd.luks.devices.<name>.fido2` already exists.

flokli replied:
I'd prefer if disabling these kernel modules would an opt-in config option.

We can't determine at evaluation time if there's a fido luks slot configured.

ajs124 replied:
fair enough. iirc we _can_ determine at eval time if the kernel even has these modules enabled. maybe using `system.requiredKernelConfig`?

ajs124 replied:
looking at the kernel config, these options/modules don't even seem to be available on non-x86 systems or systems without ACPI, so I'm really not sure they should be enabled by default

flokli replied:
Hmmh, that'd be a problem. Can we check if the kernel has these modules configured, and only include them in that case?

ajs124 replied:
I think something like this might work:
```diff
--- a/nixos/modules/system/boot/systemd/initrd.nix                                                                                                                                    
+++ b/nixos/modules/system/boot/systemd/initrd.nix
@@ -334,8 +334,11 @@ in {
               
     boot.initrd.availableKernelModules = [
       "autofs4"           # systemd needs this for some features
-      "tpm-tis" "tpm-crb" # systemd-cryptenroll              
-    ];                                                                                    
+    ]                                                                                     
+      # systemd-cryptenroll                                                                                                                                                          
+      ++ lib.optional (config.boot.kernelPackages.kernel.config.isEnabled "TCG_TIS") "tpm-tis"
+      ++ lib.optional (config.boot.kernelPackages.kernel.config.isEnabled "TCG_CRB") "tpm-crb"                                                                     
+    ;                                                                                     
                                
     boot.initrd.systemd = {                                                               
       initrdBin = [pkgs.bash pkgs.coreutils cfg.package.kmod cfg.package] ++ config.system.fsPackages;                                                                              
```
but only if we actually set these options and don't rely on kernel defaults. Maybe.

For some reason `nixosTests.systemd-initrd-luks-fido2` and `nixosTests.systemd-initrd-luks-tpm2` build even with:
```
     boot.initrd.availableKernelModules = [
       "autofs4"           # systemd needs this for some features
     ];
```
so I'm not sure how to test.

zhaofengli replied:
The VM test passes without adding any extra modules but on (some?) actual hardware it doesn't.
